Content.Server/StoreDiscount/Systems/StoreDiscountSystem.cs

```diff
@@ -0,0 +1,231 @@
+        foreach (var (discountCategory, itemsCount) in chosenDiscounts)
+        {
+            if (itemsCount == 0)
+            {
+                continue;
+            }
+
+            if (!listingsByDiscountCategory.TryGetValue(discountCategory, out var itemsForDiscount))
+            {
+                continue;
+            }
+
+            var chosen = _random.GetItems(itemsForDiscount, itemsCount, allowDuplicates: false);
+            foreach (var listingData in chosen)
+            {
+                var cost = listingData.Cost;
+                var discountAmountByCurrencyId = new Dictionary<string, FixedPoint2>();
+                foreach (var kvp in cost)
+                {
+                    if (!listingData.DiscountDownTo.TryGetValue(kvp.Key, out var discountUntilValue))
+                    {
+                        continue;
+                    }
+
+                    var discountUntilRolledValue =
+                        _random.NextDouble(discountUntilValue.Double(), kvp.Value.Double());
+                    var leftover = discountUntilRolledValue % 1;
+                    var discountedCost = kvp.Value - (discountUntilRolledValue - leftover);
+
+                    discountAmountByCurrencyId.Add(kvp.Key.Id, discountedCost);
+                }
+
+                var discountData = new StoreDiscountData
+                {
+                    ListingId = listingData.ID,
+                    Count = 1,
+                    DiscountAmountByCurrency = discountAmountByCurrencyId
+                };
+                list.Add(discountData);
+            }
+        }
+
+        return list.ToArray();
+    }
+
+    /// <summary>
+    /// Settings for discount initializations.
+    /// </summary>
+    public sealed class DiscountSettings
+    {
+        /// <summary>
+        /// Total count of discounts that can be attached to uplink.
+        /// </summary>
+        public int TotalAvailableDiscounts { get; set; } = 3;
+
+        /// <summary>
+        /// Maximum count of category 2 (not cheap stuff) items to be discounted.
+        /// </summary>
+        public int MaxVeryRareDiscounts { get; set; } = 1;
+
+        /// <summary>
+        /// Maximum count of category 0 (very low-costing stuff) items to be discounted.
+        /// </summary>
+        public int MaxRareDiscounts { get; set; } = 2;
+
+        /// <summary>
+        /// % chance (out of 100) to roll discount on listing item with category <see cref="DiscountCategory.RareDiscounts"/>.
+        /// Is considered only after comparing roll result with <see cref="VeryRareDiscountChancePercent"/>.
+        /// </summary>
+        public int RareDiscountChancePercent { get; set; } = 20;
+
+        /// <summary>
+        /// % chance (out of 100) to roll discount on listing item with category <see cref="DiscountCategory.VeryRareDiscounts"/>. 
+        /// </summary>
+        public int VeryRareDiscountChancePercent { get; set; } = 2;
+    }
+
```

Comment:
Baking in a bunch of categories makes no sense, just make it its own struct.

Reply:
inserted all % chances and max items to porot